lib/util/errors.js

```diff
@@ -0,0 +1,65 @@
+'use strict';
+
+const {
+  Error,
+  SafeMap,
+} = primordials;
+
+const {
+  makeErrorWithCode: _makeErrorWithCode,
+  AbortError,
+  codes: {
+    ERR_INVALID_ARG_TYPE,
+    ERR_INVALID_ERROR_MESSAGE,
+    ERR_MISSING_ERROR_MESSAGE,
+  },
+} = require('internal/errors');
+
+const {
+  DOMException
+} = internalBinding('messaging');
+
+const {
+  validateObject,
+  validateString,
+} = require('internal/validators');
+
+const {
+  isMap,
+} = require('util/types');
+
+function makeErrorWithCode(code, options = {}) {
+  validateString(code, 'code');
+  validateObject(options, 'options');
+  const {
+    Base = Error,
+    messages = new SafeMap(),
+  } = options;
+  if (typeof Base !== 'function') {
+    throw new ERR_INVALID_ARG_TYPE(
+      'options.Base',
+      'Function',
+      Base);
+  }
+  if (!isMap(messages)) {
+    throw new ERR_INVALID_ARG_TYPE(
+      'options.messages',
+      'Map',
+      messages);
+  }
+  const message = messages.get(code);
+  if (message === undefined) {
```

Comment:
We may want to add an extra check in case the code is not missing but mapped to `undefined`. It's the error path anyway.
```suggestion
  if (message === undefined && !messages.has(code)) {
```

Reply:
This is unnecessary given the `message.get(code)` immediately before. `message` will be `undefined` in either case.

Reply:
> `message` will be `undefined` in either case.

My point exactly, I don’t think we should treat a value of `undefined` the same as a missing value.

```js
const messages = new Map;
messages.set('foo', undefined);
makeErrorWithCode('foo', { messages }); // I expect a ERR_INVALID_ERROR_MESSAGE error.
makeErrorWithCode('bar', { messages }); // I expect a ERR_MISSING_ERROR_MESSAGE error.
```

It would be nice to add a test case for this BTW.

Reply:
> My point exactly, I don’t think we should treat a value of undefined the same as a missing value.

Hmmm... I don't really agree with that. I will add a test tho :-)